### DISCUSSION TITLE
Idea: Date and Time Parsing

### DISCUSSION THREAD

**Discussion by cmuell89** [explosion/spaCy#513] [New Features & Project Ideas]
@honnibal congrats on the release milestone!

I'm posting to see what sort of community exists amonst SpaCy users for more robust date and time parsing. For many NL-based applications, date and time parsing is tremendously useful but is a difficult task for a  statistical parser to provide consistent results from application to application. There are so many arbitrary dates, ranges, and holidays that can confound a purely statistical date time parser. There is potential within SpaCy's existing features (Matcher and PhraseMatcher classes) to build out a rule-based parser similar to that of Stanford's [SUTime](http://nlp.stanford.edu/software/sutime.shtml) (see SUTime [in action](http://nlp.stanford.edu:8080/sutime/process)) or wit.ai's open source [Duckling](https://duckling.wit.ai/). 

SUTime parses text using  token-oriented regex patterns to map to semantic objects representing dates, duration, ranges etc,. These semantic objects can be combined to create higher-order date and time annotations based on additional rules. 

Duckling takes a more functional approach (written in Clojure) and uses primitive rule based functions that can be composed into higher order functions, ultimately returning potential date and time annotations. I believe it further uses a Naive Bayes classifier to pick out the likeliest annotation as Duckling can also parse out any sort of rule-based annotation (currency etc,.) thereby resolving conflicts.

I am curious if there is  a market for creating a similar parser built within the SpaCy ecosystem (and under the same license of course). Some requirements would be as follows:
- Create a DSL to represent rules and patterns.
- Recognize what  usable features in SpaCy can be extended #486 #503 and what features must be implemented 
- Method to combine rules in to higher order representations (object composition? tree structure? function composition?)
- Documentation: Tutorials on rule creation, source code, etc,.
- Integrate parser into SpaCy's pipeline.
- Make it FAST, in the spirit of SpaCy

If this is not the right forum for this discussion please let me know. (Is SpaCy's gitter up a running?). I'd be happy to maintain this project's development but it would be a significant undertaking and a bit beyond my level of expertise.

Cheers!

Carl

**Comment by aborsu** [comment]
I have a pretty good date parsing grammar that works for en_US, en_UK and fr that was written a while back using [unitex](http://www-igm.univ-mlv.fr/~unitex/) but I would be willing to convert it and open source it.

I have been trying to switch my company's nlp activities to spaCy, but since we have french and dutch nlp pipelines, we are still far away from it. This would definitely help though.

**Comment by honnibal** [comment]
@aborsu: That'd be really great!

I'm keen to have time and date parsing in spaCy. I'd love for someone to step forward to lead development on this.

Lately I've been thinking a bit about local grammars/finite-state automata. I think these things are standard and well-proven. It'd be good to have an implementation of this, especially if we support a standard format such as unitex.

**Comment by cmuell89** [comment]
@aborsu That sounds wonderful. Might you elaborate more on what you meant by converting what you have written or used from unitex for use in SpaCy? I am not very familiar with unitex but I am assuming you patched together some of the underlying C/C++ modules for your purposes rather than make use of the GUI. I would be glad to help out if you need.

@honnibal Do you have any reading recommendations on how FSAs are generally implemented for use in something like date time parsing? What sort of states are represented, what constitutes a transition to a new state etc,. Do you envision a custom implementation or do you think an external library might suit our needs?

What do you see as the major steps needed to bring a date and time parsing into SpaCy?

Just some questions to get ideas brewing.

**Comment by aborsu** [comment]
Unitex is actually written in c++. The gui uses jni bindings (very well done jni bindings may I add) to direct the c++ code. Unitex is actually a lot more than just finite state automata's they also support using transformative grammars, cascades of transformative grammars, probabilistic taggers (but they don't provide any) and other stuff...
Since we don't currently have the ability to train our own models for lack of data, we mostly use it as a date and event extractors using hand written rules.
I really recommend reading the manual of Unitex as it is really powerful for pattern matching.

Patterns in unitex can be reprsented as graphs and look like [this](http://www-igm.univ-mlv.fr/~unitex/graph.png). They are read left to right and the top path is preferred in case of ambiguity. Graphs created using unitex are saved as .grf files which look like this.

```
#Unigraph
SIZE 1188 840
FONT Times New Roman:  10
OFONT Arial Unicode MS:B 12
BCOLOR 16777215
FCOLOR 0
ACOLOR 13487565
SCOLOR 16711680
CCOLOR 255
DBOXES y
DFRAME y
DDATE y
DFILE y
DDIR n
DRIG n
DRST n
FITS 100
PORIENT L
#
8
"<E>" 70 203 1 5 
"" 624 203 0 
"<!DIC>" 209 203 1 3 
"road+street+way+avenue" 284 203 1 6 
"<E>/{,street=$NAME$.LOC}" 429 203 1 1 
"$NAME(" 134 203 2 2 7 
"$NAME)" 354 203 1 4 
"upper+lower" 154 203 1 2 
```

The top part contains information about how to display the graph, the bottom part contains the graph. This graph detects street names by detecting a word that is not in our dictionnary and followed by road OR street OR way OR avenue and it outputs this `,street=UNKNOWN_WORD.LOC`.

If I run unitex and tell it to output a concordancer, it will give me a list of spans output so for example ->
`17 26    ,street=Berthelot.LOC`

The grf files can link one another and can be compiled to a binary format used in unitex.

To incorporate this in spacy would require either rewriting the code in spaCy to be compatible with the grf or bin files used in unitex. Or wrapping around unitex.

**Comment by honnibal** [comment]
There's a Python wrapper for unitex here: https://pypi.python.org/pypi/pyunitex

We should check out whether this is good enough, and if it's not, write our own wrapper using Cython. The advantage of the Cython wrapper is that we can call into the C++ code with no performance penalties --- this might be appealing for some workflows.

Unitex is LGPL licensed, so we can't have it as an installation requirement of spaCy. Instead, we could have it as a capability requirement, so you do `pip install spacy[unitex]`. Some organisations have different policies about using MIT and LGPL licensed code, and we can't stay MIT licensed if we have an LGPL installation requirement.

**Comment by aborsu** [comment]
The wrapper hasn't been maintained in a long time. ALso to work properly, we would have to share the dictionnary the tokens and the token features which coupled to the licensing issue tell me that it might be better to inspire ourselves from unitex than to wrap it. (But I'm willing to admit that i don't have much time to do either... 😁) (I don't even really have time to do much nlp)

**Comment by cmuell89** [comment]
@honnibal Is date time parsing something you would like to incorporate into SpaCy under the same license? Do you think it is worth the development time or would you be satisfied with a capability requirement? I guess another potential cost would be the context switch from SpaCy to unitex. To make such a capability useful, a wrapper would have to ensure sensible mappings between annotations created by unitex and SpaCy (doc, span, token, lexemes, etc,.). And if we're learning towards capability requirement it might be worth looking into other candidates (although unitex appears to be a very good one the more I look into it).

@aborsu and @honnibal I'd certainly like to assist in leading any sort of implementation (in-house or external capability) but would likely need some help/guidance from those with more expertise than myself. Capable and customizable date, time, and temporal expression parsing will be very valuable for users of SpaCy, including myself, which is why I am willing to commit time to its development.

**Comment by cmuell89** [comment]
In case anyone is interested, I've come across a [paper](http://www.sciencedirect.com/science/article/pii/S0304397503005371) that describes CasSys, a software system uses Finite State Transducer cascades for NER and various other tasks. It was developed using the Intex tool set which is the close-sourced precursor to Unitex.

**Comment by honnibal** [comment]
Let's have a call to discuss this. Carl, are you free Friday morning (PDT)? Alternatively, if this can wait until next Wednesday, I'll be able to do a bit more reading, and be better prepared.

**Comment by cmuell89** [comment]
Yes, I am free for a call either day. Let's shoot for next Wednesday, the 12th. As you guessed, I am in California (PDT) but I am flexible to waking up early or staying up late if it works best for you.

**Comment by cmuell89** [comment]
Additional resources referenced in [this](http://openfst.org/twiki/bin/view/FST/FstHltTutorial) tutorial for OpenFST provide good background and motivation on FST for application in language. All still quite new to me so I have quite a lot of reading to do.

**Comment by cmuell89** [comment]
@honnibal Any update on having a discussion this Wednesday? Cheers.

**Comment by viksit** [comment]
Incidentally one of the best date parsers out there is happens to be https://github.com/bear/parsedatetime. Might be worth incorporating with a nice spacy API. It uses the Apache license.

**Comment by cmuell89** [comment]
I have toyed with this library a bit. Correct me if I am wrong but I found it effective at parsing expressions but not at extract those expressions from text. It may parse "last week" into a date and time annotation but does not parse "last week" out of "We met for dinner and a movie last week."

**Comment by rmdort** [comment]
I actually wrote a date parser for chatterbot.

https://github.com/gunthercox/ChatterBot/blob/master/chatterbot/utils/parsing.py

If interested, i can port it to spacy.

**Comment by cmuell89** [comment]
Your approach seems similar to that of the rule based pattern matching parsers. Starting with basic core patterns, compile larger and larger patterns that meet a specific pattern group.

I think if we're going to port something like this to SpaCy, we should think how to best present the structure of patterns and regex match rules such that any user of SpaCy can easily extend the rules. Addtionally it should mesh into SpaCy's annotation schema.

I'll definitely check it out!

**Comment by cmuell89** [comment]
@rmdort 
Ran into a litter error giving your datetime parser a test run. I get the following error:

```
Traceback (most recent call last):
File "DatetimeParsing/parser.py", line 674, in <module> print(datetime_parsing(query))
File "DatetimeParsing/parser.py", line 658, in datetime_parsing matches.append((m.group(), fn(m, base_date), m.span())
File "DatetimeParsing/parser.py", line 203, in <lambda> m.group('convention')
TypeError: unsupported operand type(s) for +: 'NoneType' and 'datetime.timedelta'
```

Which is occuring this this code block: 

``` pyhton
   (re.compile(
        r'''
        (?P<time>%s) # this, next, following, previous, last
        \s+
        (?P<dow>%s) # mon - fri
        ((\s|,\s|\s(%s))?\s*(%s))?
        '''% (re_timeframe, day_names, re_separator, re_time),
        (re.VERBOSE | re.IGNORECASE),
        ),
        lambda m, base_date: dateFromRelativeDay(
            base_date,
            m.group('time'),
            m.group('dow')
        ) + timedelta(**convertTimetoHourMinute(
                m.group('hour'),
                m.group('minute'),
                m.group('convention')
            ))
    ),
```

When testing the datetime_parsing() method with the test text "this month" inside a simple prompt loop. The parser.py file an exact match of your parsing file with a little console input loop to test expressions. I cannot seem to find a re.compile() statement for those types of expressions but I'm still exploring.

Cheers,
C

**Comment by rmdort** [comment]
Oh i missed to capture month.

Pushed an update here: https://github.com/rmdort/ChatterBot/blob/timeparser/chatterbot/utils/parsing.py

**Comment by robinsonkwame** [comment]
(Stumbled upon this issue while looking for similar functionality ...)
@cmuell89 from reading [parsing natural language with date parser](https://blog.scrapinghub.com/2015/11/09/parse-natural-language-dates-with-dateparser/) I wonder if a strategy like this may be worth a look:

a) Use spaCy POS/NER tagging to identify high confidence candidate date/time sequences
b) Feed high confidence data/time sequences to `dateparser`, yield successful conversions

**Comment by cmuell89** [comment]
@robinsonkwame I think that could be an effective strategy. Although the approach is entirely dependent on the performance of spaCy's ability to recognize date time phrases. This ability (to no fault of the author ;)), currently struggles a bit in terms of coverage of date time expressions. As its NER is trained on BILOU labeled expressions my guess is many date time phrases have likely not been introduced into the training data. I think your suggested approach could work really well if we have an effective phrase parsing mechanism. By passing candidate expressions or expression patterns into some formalized mechanism that utilizes DateParser we can make use of the common formatting returned by the library.

I've been sidetracked by other work for my project but I do hope to revisit datetime parsing in a big way in the coming month or two. @rmdort 's library could be a good start for more hand-written rule parsing (regex, token expressions) that we could then pass into dateparse. Or we could look to make something original to spaCy by use of Finite State Automata and CFG approaches.

**Comment by robinsonkwame** [comment]
@cmuell89 Those are great points. I agree that for candidate expressions we'd want to focus on extremely high recall and tolerable precision/accuracy (70%+) and labelling time/date expressions for spaCy may be the way to do this. Another thought is that since date/time expressions aren't that common it's probably faster to accumulate potential phrases and pass them to the dedicated phrase parsing mechanism, rather than passing in the entire document.

Regarding date/time libraries of interest, the [Natty Date parser](http://natty.joestelmach.com/) (@joestelmach) might be of interest. It's quite good and is MIT licensed (see: https://github.com/joestelmach/natty and http://natty.joestelmach.com/try.jsp). The big catch is that it's written in Java ... :(

**Comment by mtwilliams** [comment]
Also a part of the Java ecosystem, but of interest as well is [Duckling](https://duckling.wit.ai).

**Comment by robinsonkwame** [comment]
@cmuell89  I'm torn between using @rmdort library versus making something original to spaCy ([training ](https://github.com/explosion/spaCy/issues/187) a new entity, `natural language time expression`). A large issue is training data ... new natural language time/date snippets can be generated (ironically enough) by using regular expressions as a guide. But if that's the case then why not just use the regexes in the first place? I supposed it comes down to a trade off between development time, speed and contributing a new entity. Thoughts?

**Comment by cmuell89** [comment]
@robinsonkwame I think the approach of using hand written rules and regex like functionality is likely the best one considering the lack of a formal set of date-time expressions. Additionally, date time expressions often are (and require) an exactness that other parsing algorithms might fail to provide. I am looking into using Spacy's Matcher and PhraseMatcher for potential regex like functionality that enables the creation primitive and higher order datetime patterns / rules. These features might work in conjunction with something like @rmdort 's regex styled parser.

@honnibal According to the matcher documentation, the matcher iterates over all patterns and only acts upon "on_match" callback functions after finishing this iteration. This seems to prohibit something like the following:

```
matcher.add_entity(
    "DayOfWeek", # Entity ID -- Helps you act on the match.
    {"ent_type": "DOW"} # Arbitrary attributes (optional)
)
matcher.add_pattern(
    "DayOfWeek", # Entity ID -- Created if doesn't exist.
        [ # The pattern is a list of *Token Specifiers*.
            { # This Token Specifier matches tokens whose orth field is "Monday"
              ORTH: "Monday"
            }
        ],
    label='monday' # Can associate a label to the pattern-match, to handle it better.
)
matcher.add_pattern(
    "DayOfWeek", # Entity ID -- Created if doesn't exist.
        [
            { 
              ORTH: "monday"
            }
        ],
    label='monday'
)
matcher.add_pattern(
    "DayOfWeek", 
        [ 
            { 
              ORTH: "mon"
            }
        ],
    label='monday' 
)

matcher.add_entity(
    "DateExpression",
    {"ent_type": "DE"}
)

matcher.add_pattern(
    "DateExpression",
    [
        {
            ORTH: "this"
        },
        {
            ENT_TYPE: "DayOfWeek" #### This assumes that DaysOfWeek Entities have been assigned to the ent_type of the token view of the doc.
        }
    ],
    label="DateExpression"
)
```

The idea is to create a set of core entities and patterns (days of week, months, etc,.) upon which higher order entities and patterns are built (similar to that of Stanford's SUTime and its TokenRegex feature). Am I missing something or is the matcher not capable of performing this sort of task? How about the phrase matcher?

I would be willing to build a set of handwritten rules for many date time expressions if this functionality exists in the Matcher classes.

**Comment by honnibal** [comment]
I figured rule cascades could be implemented by having the earlier patterns set boolean flags, that the later patterns reference. If the flags are decideable over single words, you can simply do this at the vocab level instead of using the matcher.

**Comment by robinsonkwame** [comment]
@cmuell89 Okay, sounds reasonable, no FSA/CFG and just regex.

Would you welcome help on this? I've identified a great set of labeled natural date time expressions for testing/validation and am happy to help with rigorous testing and/or implementation as appropriate.

**Comment by cmuell89** [comment]
I am currently exploring the Matcher's functionality to assess its feasibility for something as complex as date time expressions and cascading rules. Currently, I am creating a basic text to int parser using SpaCy's tagger and matcher to identify certain text-based numerical expressions. The idea is to convert 'one thousand three hundred twenty' to 1320.

@honnibal In the Matcher tutorial example, you have the following:

`IS_TITLE_TERM = matcher.vocab.add_flag(lambda string: string.lower() in titles)`

In my prototype I make use of this same code to create two flags via the following:

```
PrimitiveNumbers = {
    'zero': 0,
    'one': 1,
    'two': 2,
    'three': 3,
    'four': 4,
    'five': 5,
    'six': 6,
    'seven': 7,
    'eight': 8,
    'nine': 9,
    'ten': 10,
    'eleven': 11,
    'twelve': 12,
    'thirteen': 13,
    'fourteen': 14,
    'fifteen': 15,
    'sixteen': 16,
    'seventeen': 17,
    'eighteen': 18,
    'nineteen': 19,
    'twenty': 20,
    'thirty': 30,
    'forty': 40,
    'fifty': 50,
    'sixty': 60,
    'seventy': 70,
    'eighty': 80,
    'ninety': 90,
    'hundred': 100
}

Magnitudes = {
    'thousand': 1000,
    'million': 1000000,
    'billion': 1000000000
}

....

 def matcherBuilder(self):
       # Primitive Numbers are numerical expressions such as zero, one, two, three, four etc,. that map to integer values in a dict.
        PRIMITIVE_NUMBER = self.matcher.vocab.add_flag(lambda string: string.lower() in PrimitiveNumbers.keys())

        # Magnitudes are numerical expressions such as thousand, hundred etc,.
        MAGNITUDES = self.matcher.vocab.add_flag(lambda string: string.lower() in Magnitudes.keys())
```
The following examples are two continuations of the of the matcherBuilder function. The first example produces matches as expected. The second produced unexpected match results (perhaps due to my misunderstanding the Matcher's behavior).


First example:
```
         self.nlp.matcher.add_entity('Primitive_Group')
                 self.nlp.matcher.add_pattern('Primitive_Group', [{PRIMITIVE_NUMBER: True'}])
                 doc = self.nlp(u"one million eight hundred thousand twenty seven")
                 for ent_id, label, start, end in self.nlp.matcher(doc):
                     print(ent_id)
                     print(doc[start:end])

# 1510404
# one
# 1510404
# eight
# 1510404
# hundred
# 1510404
# twenty
# 1510404
# seven
```

Second example. 
It appears that MAGNITUDES are being included but the below pattern should be matching one or more tokens where PRIMITIVE_NUMBER is True.

```
        self.nlp.matcher.add_entity('Primitive_Group')
                self.nlp.matcher.add_pattern('Primitive_Group', [{PRIMITIVE_NUMBER: True', 'OP':'+'}])
                doc = self.nlp(u"one million eight hundred thousand twenty seven")
                for ent_id, label, start, end in self.nlp.matcher(doc):
                     print(ent_id)
                     print(doc[start:end])

# 1510404
# one million
# 1510404
# eight hundred thousand
# 1510404
# hundred thousand
```

Am I incorrectly using the quantifiers or the 'add_flag()' method of vocab?

**Comment by cmuell89** [comment]
@honnibal How would I go about building second order patterns beyond setting word based attributes using SpaCy's builtin Lexeme flags? 

Here's sort of the functionality I was hoping to achieve through the Matcher and perhaps it would require an expansion of features. I envision the Matcher to have functionality similar to that of TokensRegex (Used for Stanford's CoreNLP and SUTime). 

Text:
Set an appointment this Monday at 9:30 PM

```
DAY_OF_WEEK_FLAG = self.matcher.vocab.add_flag(lambda string: string.lower() in DayOfWeekArray)
matcher.addEntity('DateExpression')
matcher.addEntity('SpecificTimeExpression')
matcher.addPattern('DateExpression', [{ORTH: 'this'},{DAY_OF_WEEK_FLAG: TRUE}])
matcher.addPattern('SpecificTimeExpression', [{ORTH: 'at'}, {ORTH: '9:30'}, {ORTH: 'PM'}]

# Matches:
# this Monday
# at 9:30 PM

matcher.addAttributePattern('DateTimeExpression', [{ATTRIBUTE: DateExpression}, {ATTRIBUTE: SpecificTimeExpression}])

# Matches:
# this Monday at 9:30 PM
```

My guess is that Attributes would require a new struct or class and perhaps an additional state machine to parse over higher order matches within the __call__ method of the Matcher class.

```
cdef struct AttributePatternC:
    cdef vector[TokenPatternC*] patterns
    str attr_name
    quantifier_t quantifier
```

Additionally, it would be neat to incorporate regex based pattern requirements. This way, the above pattern for SpecificTimeExpression could be rewritten as:

```
matcher.addPattern('SpecificTimeExpression', [{ORTH: 'at'}, {REGEX: r"([0-9]?[1-9]):([0-9][0-9])"'}]
```

Perhaps this should be done through the acceptor callback? Although from what I gather, the acceptor callback act on matches found using patterns for that entity and provide proceeding functionality to act on the matches. Acceptor functions are useful in assigning labels and tags etc, but there is no real support for identifying patterns (or as a requirement for a pattern) based on regular expressions, correct?

**Comment by honnibal** [comment]
Okay so if I'm understanding correctly, you want to reference multi-word labels in the patterns, to build up more abstract grammars.

You could merge the tokens in each subexpression, so that you have a single labelled token to operate on. Then you can use another pattern based on that.

Alternatively you could use IOB-style tags to mark the beginnings and ends of the entities you're interested in. Then you can use token patterns, with a one-or-more quantifier on the "I" tag. This will potentially get ugly though...You'll probably want multiple tags, and you'll only have one slot for them. I think the entity type is stored as an int32. So if you only have 32 entity types, you could store multiple per token by using the `.ent_type` attribute as a bitfield.

**Comment by zeluspudding** [comment]
Sooooo... how can we parse dates and times using spaCy?

**Comment by evanmiltenburg** [comment]
Nobody seems to have mentioned [HeidelTime](https://github.com/HeidelTime/heideltime) yet. As far as I know, this is one of the best tools for getting times and dates in many different languages.

**Comment by ines** [comment]
Quick update: This might be a nice use case for the new [custom processing pipeline components](https://alpha.spacy.io/usage/processing-pipelines#custom-components) and [extension attributes](https://spacy.io/usage/processing-pipelines#custom-components-attributes) introduced in v2.0!

**Comment by DuyguA** [comment]
Date time format is a tricky question by the way, which sort of strings one wants to recognize.

For instance, my company does sales and customer care. We have CFG parsers both available in English and German. We recognize also relative time strings:

```
- yesterday around 15
- in 6 months
- after Christmas vacation
- 2 March Mon., 15.00
- tomorrow 12.00
- this Friday, after 12.00
```

Here are some example strings that are recognized in German:

```
seit ein paar Jahren
seit Februar
seit mehr als 10 Jahren
seit bald 4 Jahren
seit September letztes jahres
seit Mitte 2010
seit vielen Jahren

Ende September
im Januar
ende Februar/Anfang Maerz
ab Maerz nachsten Jahres
im Sommer
Sommer 2016
2. Quartal 2016
dritten Quartal 2016
vierten Quartal
IV. Quartal 2015
am Donnerstag um 10.00Uhr
Dienstag den 22.3. von 08:30 - 10:30 und 13:00-15:00Uhr
Morgen oder Freitag um 9Uhr
Di., den 29.12. oder Mi, den 30.12
freitag Nachmittag zwischen 1530 und 17Uhr
Freitag 15:30 oder 16:00 Uhr
```

... it really depends what sort of strings one wants to recognize. For instance, we recognize Quartals because of we work in business domain.

Both German and English parsers are CFG parsers, hence lexical and language dependent. We use Lark and pyparsing. 

@ines , if you're interested I'll speak to my boss and we can provide the parsers. We'll be happy to provide:)

**Comment by ines** [comment]
@DuyguA This looks great – and yes, this is definitely interesting! 👍 

It'd be really cool to have this a separate plugin that users could plug into their spaCy pipeline. I don't know if your company already publishes open-source software, but this could be nice opportunity – you get to publish the parser and maintain the project, and I'm sure a lot of spaCy users would really appreciate it. We have a lot of users working with business text in similar domains, so your plugin could definitely be very popular!

As an idea, once you've extracted the spans, the pipeline component could set additional `TIME` and `DATE` entities on the `Doc` – [see here for more info](https://spacy.io/usage/linguistic-features#setting-entities).  For example, something like this (pseudo code and all, but I hope you get the idea):

```python
from spacy.tokens import Span

parser = load_your_parser()  # load your parser somehow

def time_and_date_parser(doc):
     times = parser(doc)  # run your parser over the document
     for start, end in times:  # assuming it returns the start and end token of the span
         entity = Span(doc, start, end, label=doc.vocab.strings['TIME'])  # define span
         doc.ents = list(doc.ents) + [entity]  # add entity to doc.ents
    return doc  # return Doc object for next pipeline component
```

Usage could then look something like this:

```python
import spacy
import time_and_date_parser  # or some other cool name

nlp = spacy.load('en_core_web_sm')
nlp.add_pipe(time_and_date_parser, before='ner')  # add it before spaCy's entity recognizer

doc = nlp("How about this Friday, after 12.00?")
print([ent.text, ent.label for ent in doc.ents])
# ['this Friday, after 12.00', 'TIME']
```

spaCy's entity recognizer respects already set entities, so if the component is added before `'ner'` in the pipeline, it shouldn't interfere.

**Comment by Meekohi** [comment]
I would be extremely interested in this. Currently trying to parse stuff like "Fourth and Second Wednesday of every month, 8-9 a.m. (2018)" and am pretty stumped.

**Comment by DuyguA** [comment]
Sorry for late answer. I made a blog post to describe how to design a context free grammar:

https://duygua.github.io/blog/2018/03/28/chatbot-nlu-series-datetimeparser/

I used Lark, but one can use PyParsing and of course yacc:)

For more advanced colleagues, I also included  efficiency issues and language modelling.

**Comment by azarezade** [comment]
@DuyguA Thanks for the blog post!
But, how can someone convert the parsed string to an standard time format (given a reference time)?

**Comment by DuyguA** [comment]
One needs an attribute grammar. Basically parse tree is mapped to an assembly language. Sth like this:

https://en.m.wikipedia.org/wiki/Attribute_grammar


https://github.com/lark-parser/lark/blob/master/examples/calc.py

I can extend the blog post. This way one can convert relative dates to Python objects.

**Comment by azarezade** [comment]
Thanks, that would be nice if you can extend your blog post.

**Comment by azarezade** [comment]
Did anybody compared NLTK with lark in terms of speed and accuracy?

**Comment by DuyguA** [comment]
Yes, myself :)

Lark is the fastest, PyParsing takes the second place, NLTK is not very competent.

NLTK is not very competent in efficiency issues in general. It's not very really competent for realtime applications, NLTK looks like it's good for academical purposes. Nobody I know uses NLTK for industrial applications. Most of my text miner friends make SpaCy + Tensorflow/Keras backends (including myself :grin: )

**Comment by azarezade** [comment]
Great! 👍 
So, I will use Lark too ;)

**Comment by DuyguA** [comment]
OK, I don't want to speak too much :grin: but:

1. Importing two similar packages (in this case both spacy and nltk) is not good in the view of software engineering. 
2. Importing a big library for only one functionality (in this case nltk-its parser) is not good for efficiency as well.

If you really want to use NLTK, go for Earley parser (there exists one). We both English and German CFG for our inhouse e-mail bots and both grammars are very ambigious. I think any parser less then Earley cannot cover the language of date-time strings.

**Comment by piyushrj** [comment]
Hello, I'm following this thread to extract Dates from text corpus, I'm using Spacy's ner model to get dates and have created some custom rules using `Matcher` class to identify some missed date formats. Here is my [code,](https://gist.github.com/piyushrj/29f35793949a44e17ca0f9a08b98226c) I need to test my approach and want to see its coverage for different kinds of date formats but I don’t have the data to do so, if I could get such data for doing this coverage analysis it would be of great help.

**Comment by honnibal** [comment]
I think we can finally close this 🎉 

While we don't have time and date parsing in the library itself, in v2.1 we now have:

* Correct handling of quantifiers in the `Matcher`
* Support for matching against token extensions
* Support for matching with regular expressions
* Support for matching with comparison and set-valued predicates
* The `EntityRuler` class, to make the match dictionaries into an entity recognizer
* Entry point hooks for custom pipeline components

The entry point hooks mean that grammars for things like times and dates don't need to live in the main library --- they can be published as extension modules, with no difference to user experience. I think the matcher support we now have basically completes the core library functionality needed to do this.
